docs/design/control-plane-resilience.md

```diff
@@ -0,0 +1,189 @@
+   for achieving system resilience (replication controllers, health
+   checking, service load balancing etc) should be used in preference
+   to building a separate set of mechanisms to achieve the same thing.
+   This implies that self hosting (the kubernetes control plan on
+   kubernetes) is strongly preferred, with the caveat below.
+1. **Recovery from catastrophic failure:** The ability to quickly and
+   reliably recover a cluster from catastrophic failure is critical,
+   and should not be compromised by the above goal to self-host
+   (i.e. it goes without saying that the cluster should be quickly and
+   reliably recoverable, even if the cluster control plane is
+   broken). This implies that such catastrophic failure scenarios
+   should be carefully thought out, and the subject of regular
+   continuous integration testing, and disaster recovery exercises.
+
+## Relative Priorities
+
+1. **Recovery from catastrophic failures:** having a Kubernetes cluster, and all 
+   applications running inside it, disappear forever is the worst failure mode.  
+    1. In theory a cluster can be recovered by replaying all API calls
+       that have ever been executed against it, in order, but most
+       often that state has been lost, and/or is scattered across
+       multiple client applications or groups. So in general it is
+       probably infeasible.
+    1. In theory a cluster can also be recovered to some relatively
+       recent non-corrupt backup/snapshot of the disk(s) backing the
+       etcd cluster state.  But we have no default consistent
+       backup/snapshot, verification or restoration process.  And we
+       don't routinely test restoration, so even if we did routinely
+       perform and verify backups, we have no hard evidence that we
+       can in practise effectively recover from catastrophic cluster
+       failure or data corruption by restoring from these backups.  So
+       there's more work to be done here.
+1. **Self-healing:** Most major cloud providers provide the ability to
+   easily and automatically replace failed virtual machines within a
+   small number of minutes (e.g. GCE
+   [Auto-restart](https://cloud.google.com/compute/docs/instances/setting-instance-scheduling-options#autorestart)
+   and Managed Instance Groups,
+   AWS[ Auto-recovery](https://aws.amazon.com/blogs/aws/new-auto-recovery-for-amazon-ec2/)
+   and [Auto scaling](https://aws.amazon.com/autoscaling/) etc). This
+   can fairly trivially be used to reduce control-plane down-time due
+   to machine failure to a small number minutes per failure
+   (i.e. typically around "3 nines" availability), provided that:
+    1. cluster persistent state (i.e. etcd disks) is either:
+        1. truely persistent (i.e. remote persistent disks), or 
+        1. reconstructible (e.g. using etcd [dynamic member 
+           addition](https://github.com/coreos/etcd/blob/master/Documentation/runtime-configuration.md#add-a-new-member) 
```

Comment:
Yes, in fact a majority quorum needs to be up in the former case (e.g. 3 of 5). It's perhaps worth mentioning that remote persistent disks are not magic, and do not have 100% uptime either.  Neither GCE nor AWS provide any uptime guarantees for remote PD's.   A 3 of 5 quorum running on 99.9% (3 nines) available nodes provides  99.999% (5 nines) availability, assuming independent node failures.  For correlated failure of course, you're still screwed.  But GCE PD's only aim to give less than 4 nines protection against such correlated failures anyway, per [their SLA](https://cloud.google.com/compute/sla?hl=en).  Similarly AWS EBS volumes are ["designed for 5 nines availability"](http://aws.amazon.com/ebs/details/#availabilityanddurability), but they don't actually offer any SLA around that.  So in summary, I contend that a 3 of 5 quorum on local ephemeral disks is at least as available as one on remote "persistent" disks.